src/irr/asset/COBJMeshFileLoader.cpp

```diff
@@ -355,490 +734,377 @@ asset::SAssetBundle COBJMeshFileLoader::loadAsset(io::IReadFile* _file, const as
+#endif
+                void* dst = reinterpret_cast<uint8_t*>(imgDataBuf->getPointer()) + regions_[j].bufferOffset;
+                const void* src = reinterpret_cast<const uint8_t*>(images[i]->getBuffer()->getPointer()) + images[i]->getRegions().begin()[0].bufferOffset;
+                const size_t sz = images[i]->getImageDataSizeInBytes();
+                memcpy(dst, src, sz);
 
-	// end of file. if there's an existing material, store it
-	if ( currMaterial )
-		_ctx.Materials.push_back( currMaterial );
+                ++j;
+#ifndef _IRR_DEBUG
+            }
+#endif
+        }
+
+        //assuming all cubemap layer images are same size and same format
+        ICPUImage::SCreationParams cubemapParams = images[CMTLPipelineMetadata::SMtl::EMP_REFL_POSX]->getCreationParameters();
+        cubemapParams.arrayLayers = 6u;
+        cubemapParams.type = IImage::ET_2D;
+
+        auto cubemap = ICPUImage::create(std::move(cubemapParams));
+        auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUImage::SBufferCopy>>(regions_);
+        cubemap->setBufferAndRegions(std::move(imgDataBuf), regions);
+        //new image goes to EMP_REFL_POSX index and other ones get nulled-out
+        images[CMTLPipelineMetadata::SMtl::EMP_REFL_POSX] = std::move(cubemap);
+        for (uint32_t i = CMTLPipelineMetadata::SMtl::EMP_REFL_POSX + 1u; i < CMTLPipelineMetadata::SMtl::EMP_REFL_POSX + 6u; ++i)
+        {
+            images[i] = nullptr;
+        }
+    }
 
-	delete [] buf;
-	mtlReader->drop();
+    return images;
 }
 
+core::smart_refctd_ptr<ICPUDescriptorSet> COBJMeshFileLoader::makeDescSet(const images_set_t& _images, ICPUDescriptorSetLayout* _dsLayout)
+{
+    if (!_dsLayout)
+        return nullptr;
+
+    auto ds = core::make_smart_refctd_ptr<asset::ICPUDescriptorSet>(
+        core::smart_refctd_ptr<ICPUDescriptorSetLayout>(_dsLayout)
+    );
+    for (uint32_t i = 0u, d = 0u; i <= CMTLPipelineMetadata::SMtl::EMP_REFL_POSX; ++i)
+    {
+        if (!_images[i])
+            continue;
+
+        constexpr IImageView<ICPUImage>::E_TYPE viewType[2]{ IImageView<ICPUImage>::ET_2D, IImageView<ICPUImage>::ET_CUBE_MAP };
+        constexpr uint32_t layerCount[2]{ 1u, 6u };
+
+        const bool isCubemap = (i == CMTLPipelineMetadata::SMtl::EMP_REFL_POSX);
+
+        ICPUImageView::SCreationParams viewParams;
+        viewParams.flags = static_cast<ICPUImageView::E_CREATE_FLAGS>(0u);
+        viewParams.format = _images[i]->getCreationParameters().format;
+        viewParams.image = _images[i];
+        viewParams.viewType = viewType[isCubemap];
+        viewParams.subresourceRange.baseArrayLayer = 0u;
+        viewParams.subresourceRange.layerCount = layerCount[isCubemap];
+        viewParams.subresourceRange.baseMipLevel = 0u;
+        viewParams.subresourceRange.levelCount = 1u;
+
+        auto desc = ds->getDescriptors(d).begin();
+        desc->desc = core::make_smart_refctd_ptr<ICPUImageView>(std::move(viewParams));
```

Comment:
you should be getting views from the asset cache (not create new ones every single time!)!

Reply:
anyway why not create the descriptor set in the MTL loader?

Reply:
i guess it could, i remember i made this decision for some reason but i cant remember..

Reply:
views should definitely be cached.

Reply:
> i guess it could, i remember i made this decision for some reason but i cant remember..

ah i know why, because theres no way to return it

Reply:
slam a pointer to it in the metadata ;)

Reply:
goes into the MTL pipeline metadata, case closed.